#! /bin/sh
patch -p0 -f $* < $0
exit $?

2008-01-30  Werner Koch  <wk@g10code.com>

	* tdbio.c (tdbio_set_dbname): Also test for forward slash.
	* keydb.c (maybe_create_keyring): Take care of a missing slash.
	(maybe_create_keyring) [W32]: Also test for forward slash.


Index: g10/keydb.c
===================================================================
--- g10/keydb.c	(revision 4666)
+++ g10/keydb.c	(working copy)
@@ -1,5 +1,6 @@
 /* keydb.c - key database dispatcher
- * Copyright (C) 2001, 2002, 2003, 2004, 2005 Free Software Foundation, Inc.
+ * Copyright (C) 2001, 2002, 2003, 2004, 2005, 
+ *               2008 Free Software Foundation, Inc.
  *
  * This file is part of GnuPG.
  *
@@ -82,6 +83,7 @@
   int rc;
   mode_t oldmask;
   char *last_slash_in_filename;
+  int save_slash;
 
   /* A quick test whether the filename already exists. */
   if (!access (filename, F_OK))
@@ -98,6 +100,18 @@
      tricky auto-creation which is anyway only done for some home
      directory name patterns. */
   last_slash_in_filename = strrchr (filename, DIRSEP_C);
+#if HAVE_W32_SYSTEM
+  {
+    /* Windows may either have a slash or a backslash.  Take care of it.  */
+    char *p = strrchr (filename, '/');
+    if (!last_slash_in_filename || p > last_slash_in_filename)
+      last_slash_in_filename = p;
+  }
+#endif /*HAVE_W32_SYSTEM*/
+  if (!last_slash_in_filename)
+    return gpg_error (GPG_ERR_ENOENT);  /* No slash at all - should
+                                           not happen though.  */
+  save_slash = *last_slash_in_filename;
   *last_slash_in_filename = 0;
   if (access(filename, F_OK))
     { 
@@ -111,13 +125,12 @@
       if (access (filename, F_OK))
         {
           rc = gpg_error_from_syserror ();
-          *last_slash_in_filename = DIRSEP_C;
+          *last_slash_in_filename = save_slash;
           goto leave;
         }
     }
-  *last_slash_in_filename = DIRSEP_C;
+  *last_slash_in_filename = save_slash;
 
-
   /* To avoid races with other instances of gpg trying to create or
      update the keyring (it is removed during an update for a short
      time), we do the next stuff in a locked state. */
Index: g10/tdbio.c
===================================================================
--- g10/tdbio.c	(revision 4666)
+++ g10/tdbio.c	(working copy)
@@ -505,14 +505,25 @@
 	    int rc;
 	    char *p = strrchr( fname, DIRSEP_C );
 	    mode_t oldmask;
+            int save_slash;
 
-	    assert(p);
+#if HAVE_W32_SYSTEM
+            {
+              /* Windows may either have a slash or a backslash.  Take
+                 care of it.  */
+              char *pp = strrchr (fname, '/');
+              if (!p || pp > p)
+                p = pp;
+            }
+#endif /*HAVE_W32_SYSTEM*/
+	    assert (p);
+            save_slash = *p;
 	    *p = 0;
 	    if( access( fname, F_OK ) ) {
 		try_make_homedir( fname );
 		log_fatal( _("%s: directory does not exist!\n"), fname );
 	    }
-	    *p = DIRSEP_C;
+	    *p = save_slash;
 
 	    xfree(db_name);
 	    db_name = fname;
